### ISSUE TITLE
Code is complicated and confused

### ISSUE THREAD

**Issue by joefresna** [rainazakir/argos3-kilogrid#10]
https://github.com/rainazakir/argos3-kilogrid/blob/5b966a9dd2a49a1097fc2cbbf2f1d72ec9c2ce94/ARGoS_simulation/behaviours/agent_main.c#L322

All this code
```
/*-----------------------------------------------------------------------------------------------*/
/* This function is used to convert the tiles perceived in an expl-dissem cycle to quality       */
/* of  its opinion. If q>=0.5, all qr is taken as 1. The function is inverse lambda expon.       */
/*-----------------------------------------------------------------------------------------------*/

void findqualityratio(){

    qratio = ((float)tiles_of_my_option/total_tiles_found);  //find % of tiles of opinion bot supports

    if(qratio >= 0.5){ //if % more than or equal to 0.5
        qratio = 1/(((float)1/1)*1300); //like valentini model inverse lambda, 1300 instead of 1000 to increase dissem time

    }else{ //otherwise calculate dissem time based on % found 0-0.4999
        qratio = 1/(((float)tiles_of_my_option/total_tiles_found)*1300); //like valentini model inverse lambda, 1300 instead of 1000 to increase dissem time

    }

    // qratio = 1/(((float)tiles_of_my_option/total_tiles_found)*1000); //like valentini model inverse lambda

    //  printf("%d tile my op %d total tiles and qr %f \n", tiles_of_my_option,total_tiles_found, qratio);

}

/*-----------------------------------------------------------------------------------------------*/
/* This function is used to find the time to disseminate based on the output of qr from          */
/*                      findqualityratio() function.                                            */
/*-----------------------------------------------------------------------------------------------*/
void calculatedissemtime(){

    if (isnan(qratio) || isinf(qratio)){  //if quality was found to be 0
        if(MODEL == 1){ //if cross-inhibition

            if (currentopinion == UNCOMMITTED){ //if bot is uncommitted
                printf("comes to time of uncommitted agent in noisy switch \n");
                timer =  ran_expo(quncommitted); //set time to be in dissem state but will not talk
            }else{
                timer = 0;
                //timer =  ran_expo(q0);
            }
        }else{ //if not uncommitted and tiles own opinion found to be = 0
            //printf("% f  goes directly to voting    \n", qratio);
            timer = 0;  //no time as will go directly to poll/noisyswitch state
            //timer = ran_expo(0.003);
        }


    }else{ //if quality was not 0

        //printf("% f  goes to random  \n", qratio);
        timer = ran_expo(qratio); //get the time to disseminate based on quality
        //}
    }
}
```

can't it be summarized in a more readable version:

```
const float mean_dissemination_time = 1300.0;

void findqualityratio(){
   if (total_tiles_found > 0){
    qratio = ((float)tiles_of_my_option/total_tiles_found);
    } else {
        qratio = 0;
     }
}

 void calculatedissemtime(){

  if(MODEL == 1 && currentopinion == UNCOMMITTED){ // if MODEL is cross-inhibition and bot is uncommitted

       timer =  ran_expo(quncommitted); //set time to be in dissem state but will not talk

    } else {
       double lambda = 1.0 / ( min(1.0, qratio*2) *mean_dissemination_time);
        timer = ran_expo(lambda); 
    }
}
```

I did it quickly, so please check if I did not introduce any error / bug.

**Comment by rainazakir** [comment]
So I tried the code, however this line is altering the lambda for qratio between (0.001-0.4999) as its doubles it:
`       double lambda = 1.0 / ( min(1.0, qratio*2) *mean_dissemination_time);`

I have made changes in 895fb778f0720e2c969115979c28f8ef1aea2dd9

**Comment by joefresna** [comment]
In the current code, it seems that there is a discontinuity between 0.499999 and 0.5.

Again, as we discussed in person there are three distinct quantities:
* percentage of readings of your colour on all number of readings (let's call it r for readings)
* assigned quality to the percentage (let's call it q for quality)
* dissemination time (let's call it t_d )

So we have two functions, f() and g(), where
q = f( r )
and
t_d = f( q )

Does this make sense?
Please let me know if something is unclear.

Regarding the functional form of f(), we can have that
q = f(r) = r
therefore the quality is identical to the reading. E.g., if you read from the ground 0.3, the quality is 0.3.
Otherwise, we can also have
q = f(r) = min(1, 2*r)
(as we discussed in person) in this case there is again a linear relationship between r and q, however with double the slope. It saturates at 1 for r >= 0.5.

I do not have a strong opinion on which to select among the two. What do you think?

The equation that you used is a mix of the two equations and I find them counterintuitive...

when r < 0.5, 
f(r) = r
and when r>= 0.5
f(r) = 1

Maybe you can explain the reasoning behind it. Otherwise, we can test one of the two functions described above.


Regarding the function g(), I think we agree that it is:
t_d = g( q ) =  -log(1- u) * q * mean_dissemination_time
where u is a random number in [0,1).

Please let me know if something is unclear or if I made some mistake in my reasoning (it can be). ;-)

**Comment by rainazakir** [comment]
Thanks for the explanation, I think I understand better now and do realize I have created a mix and match of the two concepts of f(). What I wanted to achieve was the qualities <0.5 are taken as it is while >0.5 saturates at 1 cause from the discussions we had in person I assumed this is how we wanted to implement it. Now that I think of it, I think that  `q = f(r) = min(1, 2*r) ` might give us equivalent of what we wanted to do without creating the mix. I also do not have a strong opinion towards any of the two functional forms of f(), initially I did have a strong opinion towards  `q = f(r) = r` cause I thought not doing so was influencing our results which later we found was due to exploration time being too less. I think lets stick to  `q = f(r) = min(1, 2*r)` and get some results to test.
